packages/core/android/src/main/kotlin/com/datadog/reactnative/DdSdkBridgeExt.kt

```diff
@@ -6,44 +6,187 @@
-    val map = WritableNativeMap()
-    for ((k,v) in iterator()) {
+/**
+ * Converts the [Map] to a [WritableNativeMap].
+ */
+internal fun Map<*, *>.toWritableMap(): WritableMap {
+    return this.toWritableMap(
+        createWritableMap = { WritableNativeMap() },
+        createWritableArray = { WritableNativeArray() }
+    )
+}
+
+/**
+ * Converts the [Map] to a [WritableMap].
+ * @param createWritableMap a function to provide a concrete instance for WritableMap(s)
+ * @param createWritableArray a function to provide a concrete instance for WritableArray(s)
+ */
+internal fun Map<*, *>.toWritableMap(
+    createWritableMap: () -> WritableMap,
+    createWritableArray: () -> WritableArray
+): WritableMap {
+    val map = createWritableMap()
+
+    for ((k, v) in iterator()) {
         val key = (k as? String) ?: k.toString()
-        @Suppress("NotImplementedDeclaration")
         when (v) {
             null -> map.putNull(key)
             is Int -> map.putInt(key, v)
             is Long -> map.putDouble(key, v.toDouble())
             is Float -> map.putDouble(key, v.toDouble())
             is Double -> map.putDouble(key, v)
             is String -> map.putString(key, v)
-            is List<*> -> map.putArray(key, v.toWritableArray())
-            is Map<*, *> -> map.putMap(key, v.toWritableMap())
-            else -> TODO()
+            is Boolean -> map.putBoolean(key, v)
+            is List<*> -> map.putArray(
+                key,
+                v.toWritableArray(
+                    createWritableMap,
+                    createWritableArray
+                )
+            )
+            is Map<*, *> -> map.putMap(
+                key,
+                v.toWritableMap(
+                    createWritableMap,
+                    createWritableArray
+                )
+            )
+            else -> Log.e(
+                javaClass.simpleName,
+                "Unhandled type ${v.javaClass.simpleName} has been ignored"
+            )
         }
     }
+
     return map
 }
+
+/**
+ * Recursively converts the [ReadableMap] to a [Map] which only contains Kotlin stdlib objects,
+ * such as [List], [Map] and the raw types.
+ */
+internal fun ReadableMap.toMap(): Map<String, Any> {
+    val map: HashMap<String, Any> = this.toHashMap()
+    val iterator: Iterator<String> = map.keys.iterator()
+
+    while (iterator.hasNext()) {
+        val key = iterator.next()
+        try {
+            when (val type = getType(key)) {
+                ReadableType.Map -> {
+                    map[key] = getMap(key)?.toMap() ?: Log.e(
+                        javaClass.simpleName,
+                        "Cannot convert nested map for key: $key"
+                    )
+                }
+
+                ReadableType.Array -> {
+                    map[key] = getArray(key)?.toList() ?: Log.e(
+                        javaClass.simpleName,
+                        "Cannot convert nested array for key: $key"
+                    )
```

Comment:
I think that by using such expression `map[key] = whateverNullable ?: Log.e(...)` you will have `Unit` as a value for `map[key]` if `whateverNullable` is indeed `null` by some reason (if I'm not wrong).

We need either to re-write it to simple `if-else`, having:

```
val value = getMap(key)?.toMap()
if (value != null) { map[key] = value } else { Log.e(...) }  
```

Reply:
indeed, it will be `Unit` in the `map[key]` in such scenario: link to reproduce https://pl.kotl.in/b54hT6ggk

Reply:
Yes, that's a clear mistake on my side! Thank you